Gracefully handle failure of DataPtr extraction in CUDAFuture

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56518 Add tests for CUDAFuture
* #56517 Add CUDA support to a user-created torch.futures.Future
* #56516 Support custom Python classes in CUDAFuture
* #56515 Allow to specify a set of device for CUDAFuture
* #56514 Split CUDAFuture into a .h and a .cpp file
* #56513 Avoid defining RpcCUDAFuture subclass in TensorPipe agent
* #56512 Merge ivalue::Future's markCompleted and markCompletedWithDataPtrs
* **#56511 Gracefully handle failure of DataPtr extraction in CUDAFuture**
* #56510 Raise TypeErrors when IValue::getSubValues fails

CUDAFuture needs to inspect the value in order to extract DataPtrs. Sometimes it's unable to do so. So far we've handled this by raising an error when `markCompleted` is called. In this PR I'm proposing a change, which makes `markCompleted` return successfully, but instead causes the Future to be set to an error if the DataPtr extraction fails.

The advantages I see are that user code calling `markCompleted` didn't expect it to throw, and thus wasn't catching and handle that error. Which in the best case could lead to a crash, and in the worst case could lead to the Future remaining incomplete, thus not unblocking any client waiting on it. With this change those clients would be woken up and would see the error.

Differential Revision: [D27861070](https://our.internmc.facebook.com/intern/diff/D27861070/)

lw committed:
Gracefully handle failure of DataPtr extraction in CUDAFuture

CUDAFuture needs to inspect the value in order to extract DataPtrs. Sometimes it's unable to do so. So far we've handled this by raising an error when `markCompleted` is called. In this PR I'm proposing a change, which makes `markCompleted` return successfully, but instead causes the Future to be set to an error if the DataPtr extraction fails.

The advantages I see are that user code calling `markCompleted` didn't expect it to throw, and thus wasn't catching and handle that error. Which in the best case could lead to a crash, and in the worst case could lead to the Future remaining incomplete, thus not unblocking any client waiting on it. With this change those clients would be woken up and would see the error.

Differential Revision: [D27861070](https://our.internmc.facebook.com/intern/diff/D27861070/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -382,7 +382,13 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
+    try {
+      preMarkCompletedHook(value, std::move(data_ptrs));
+    } catch (const std::exception&) {
+      setErrorInternal(std::current_exception(), lock);
```
This is clearly better then the current version, but does this mean chained Futures will never be marked as completed, even if they are not going to consume the value from the root future?

lw replied:
Are you talking about child futures in the sense of those returned by a call to `parent_future.then(cb)`? In that case, my understanding is that if the `preMarkCompletedHook` fails on `parent_future`, we call `setErrorInternal` on `parent_future`, which should still invoke all its callbacks, including the one for `child_future`. Those callbacks receive `parent_future` as argument, even if `parent_future` is in an error state, thus if those callbacks try to access `parent_future`'s value an exception will be thrown, which will cause `child_future` to also be set to error. If the callback never accesses `parent_future`'s value then it will just return a value as normal and `child_future` will be set correctly.

lw replied:
BTW I'm adding a test for this in #56518.

mrshenli reviewed (approved):
LGTM!

lw committed:
Update on "Gracefully handle failure of DataPtr extraction in CUDAFuture"


CUDAFuture needs to inspect the value in order to extract DataPtrs. Sometimes it's unable to do so. So far we've handled this by raising an error when `markCompleted` is called. In this PR I'm proposing a change, which makes `markCompleted` return successfully, but instead causes the Future to be set to an error if the DataPtr extraction fails.

The advantages I see are that user code calling `markCompleted` didn't expect it to throw, and thus wasn't catching and handle that error. Which in the best case could lead to a crash, and in the worst case could lead to the Future remaining incomplete, thus not unblocking any client waiting on it. With this change those clients would be woken up and would see the error.

Differential Revision: [D27861070](https://our.internmc.facebook.com/intern/diff/D27861070/)

[ghstack-poisoned]
